### ISSUE TITLE
1.0.0 Release

### ISSUE THREAD

**Issue by dignifiedquire** [karma-runner/karma#1406]
There were some notes, and plans about a 1.0 release. I want to use this issue for any discussion and plans around it and start a complete list of tasks to do before we can release.

There is this [google doc](https://docs.google.com/document/d/1XMg9CYTsu8NIGBFyXtQS8q0KaWdvldhjLsIcEBOitjo/edit?usp=sharing) that was started by @vojtajina and there was #944, which I will close in favor of this issue. 

Below is an incomplete list of things that need doing in my opinion.

## Organizational
- [ ] Clean up issue list (close outdated issues, respond to recent issues, and figure out which bugs are important enough to be fixed before release)
- [x] Check for any api changes that need to be done before hand
- [x] Bring all frameworks and launchers to a `1.0.0` ready state.
- [x] Release `1.0.0` versions of the frameworks and launchers.
- [x] Release `0.13.0` with the latest additions.
- [x] Figure out how to handle peer dependencies (Ref https://github.com/npm/npm/issues/6565 and https://github.com/karma-runner/karma-phantomjs-launcher/issues/71) Ref: https://github.com/karma-runner/karma-phantomjs-launcher/issues/71#issuecomment-102834299 I don't think we need to change anything about how we use peerDependencies.
- [x] Fix release and merge process
- [x] Update release process to handle proper semver and releases `> 1.0`
- [x] Improve the official support situation (#1414)
- [x] Make the canary build semver compatible (#1342)
- [x] Enable github integrations on gitter
- [x] Verify CLA process
- [x] Remove cruft from `canary` branch and bring it up to date with `master`
- [x] Remove `iojs-1` from travis (ec4fa02 and d42802ad225ca81facdef565a4a121cb806bcc8a)

## Improvements
- [x] Update Socket.io to the latest version (PR #1404)
- [x] Update all dependencies to their latest versions (PR #1410)
- [x] Update all dev dependencies to their latest versions (PR #1279)
- [ ] Quality review the documentation
- [x] Ensure support for node `0.10`, `0.12` and `io.js >= 2.0`

## Bugs
See the [1.0 Milestone](https://github.com/karma-runner/karma/issues?utf8=%E2%9C%93&q=is%3Aopen+milestone%3Av1.0+label%3A%22type%3A+bug%22)

// cc @zzo @maksimr 

**Update:** I updated the [milestone](https://github.com/karma-runner/karma/milestones/v1.0) to reflect relevant issues for this.
**Update2:** Reorganize

**Comment by zzo** [comment]
Looks good - want to add fixing the mary-poppins auto-release process.
  Mark

**Comment by Dignifiedquire** [comment]
@zzo added. Ping me when you have some time I'm happy to help getting this fixed and running (also probably good to have > 1 person being able to check it when there is ab issue)

**Comment by Dignifiedquire** [comment]
@zzo Cleaned up the cruft on canary and remove iojs-1 from travis. Will do the version bump on canary later

**Comment by zzo** [comment]
great i just updated the docs on master to mention only iojs 2.x https://github.com/karma-runner/karma/commit/1b69db1ce22c5b51f6dd003ceded7ac0708b84b2 and updated the docs on canary to be .10, .12 and iojs 2.x only https://github.com/karma-runner/karma/commit/38eb63d97d54b911e2fb9e78d9323b6b4d35e94e

**Comment by Dignifiedquire** [comment]
@zzo I built the latest docs, you can check them out [here](http://karma-runner.github.io/0.12/index.html)

**Comment by zzo** [comment]
Great what is the process for doing that?  thanks

**Comment by Dignifiedquire** [comment]
You need this [setup](https://github.com/karma-runner/karma-runner.github.com#building-the-documentation) and then run https://github.com/karma-runner/karma-runner.github.com/blob/master/sync-docs.sh which makes a couple of assumptions that you should check before hand (I need to get this documented properly..)

On another note I've started with the work for publishing `1.0.0-0` as the the new canary version and I found some issues with the current system. I suggest we move to the following structure:

* Everything gets merged into master
* If we have a feat or fix we publish a new prerelease version from master
* We have a `stable` branch which points to the latest stable release, e.g. right now `0.12.31`, and we only release new versions of that if we have critical fixes that we manually cherry-pick from master into there. 

I feel that this approach would involve a lot less magic and make things easier to understand, to follow (as an outsider) and to maintain (for us). What do you think?

**Comment by Dignifiedquire** [comment]
Most important part, I forgot when to release new "feature" versions: We take a fixed cycle, e.g. every 4 weeks we release a new stable version, bumping version numbers according to semver and the commits.

**Comment by Dignifiedquire** [comment]
@zzo Okay, ignore most of the above, trying to put this into a more understandable and complete form:

Suggested release cycle/process:

## Types of Releases

* canary, e.g. `1.0.0-pre.1`
* beta, e.g. `1.0.0-beta.1`
* stable, e.g. `1.0.0` (increase minor version if no breaking changes, major version if breaking changes)

## Release Cycle

Chrome style:

* Canary, on each `feat` or `fix` merge
* Beta every week
* Stable every 6 weeks (so five beta versions and then one stable release)


## Git Management

* master -> all development
* everything else is a git tag

Does that make sense to you? What does everyone else think?

// cc @karma-runner/contributors

**Comment by lygstate** [comment]
iojs still not supported

**Comment by lygstate** [comment]
The master doesn't build with nodejs 2.1.0 because of socket.io.

**Comment by Dignifiedquire** [comment]
@zzo I've given up @karmarunnerbot for now and do the releases manually..

**Comment by zzo** [comment]
What's been happening?  Bad npm credentials still?

**Comment by Dignifiedquire** [comment]
No idea :( just not happening anything after it gets send to presubmit, and additionally travis takes > 6h at the moment to process our builds...

**Comment by Dignifiedquire** [comment]
(just merged the stuff that was piling up manually, and will do a release in the next hour or two out of it)

**Comment by dignifiedquire** [comment]
@zzo going to actually do this :)

**Comment by zzo** [comment]
what could go wrong :)  what about getting to 1.0 on frameworks & launchers?

**Comment by dignifiedquire** [comment]
Will bump them one by one over the next weeks as I get to it.

**Comment by zzo** [comment]
we can split the list and I can do some as well

**Comment by dignifiedquire** [comment]
That'd be great, let's get a list in here with checkmarks

**Comment by dignifiedquire** [comment]
We need to remember to bump the karma dependency to allow for 1.0 when updating them

**Comment by zzo** [comment]
sounds good - yep

**Comment by zzo** [comment]
Here's a start:

- [ ] karma-chrome-launcher
- [ ] karma-mocha
- [ ] karma-junit-reporter
- [ ] grunt-karma
- [ ] karma-sauce-launcher
- [ ] karma-html2js-preprocessor
- [ ] karma-requirejs
- [ ] karma-teamcity-reporter
- [ ] karma-jasmine
- [ ] karma-ie-launcher
- [ ] karma-intellij
- [ ] karma-coverage
- [ ] karma-script-launcher
- [ ] karma-slimerjs-launcher
- [ ] gulp-karma
- [ ] karma-browserstack-launcher
- [ ] karma-phantomjs-launcher
- [ ] karma-ng-html2js-preprocessor

**Comment by zzo** [comment]
actually there are other karma repos not owned by karma-runner which will probably need to be updated too - these are just the ones owned by karma-runner

**Comment by zzo** [comment]
here is script I want to use to bump version numbers of packages: 
```sh
#!/bin/bash
# Use like:  $0 karma-mocha
URL=https://github.com/karma-runner/$1.git
git clone $URL
cd $1
npm install
git remote add upstream $URL
node ../updater.js $1
git commit -a -m 'version bump for 1.0'
git push origin master
grunt release
```

'updater.js' reads package.json - moves the version number to '1.0.0' and changes all 'dependencies' and 'devDependencies' with the word 'karma' from something like '~0.2.4' to '1.x || ~0.2.4' (ignoring any git url deps like ' "shared-karma-files": "git://github.com/karma-runner/shared-karma-files.git#82ae8d02")

What do you think?

Here's 'updater.js':

```js
var fs = require('fs');
var pjson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
pjson.version = '1.0.0';
pjson = doDeps(pjson, 'dependencies')
pjson = doDeps(pjson, 'devDependencies')
fs.writeFileSync('package.json', JSON.stringify(pjson, null, '  '));

function doDeps(pjson, hashKey) {
  if (pjson[hashKey]) {
    Object.keys(pjson[hashKey]).forEach(function(key) {
      if (key.match(/karma/) && !pjson[hashKey][key].match(/^git/)) {
        pjson[hashKey][key] = '1.x || ' + pjson[hashKey][key];
      }
    });
  }
  return pjson;
}
```

**Comment by zzo** [comment]
and then once everything is at 1.0 we can go back and remove the '|| <latest previous version>' part

**Comment by zzo** [comment]
just ran it on karma-mocha - looks good - will start all the other ones now

**Comment by dignifiedquire** [comment]
@zzo thanks for getting this started, I didn't have the time for this as I hoped. There is one issue I saw with your releases, that is that you are releasing `1.0.1` and skipping `1.0.0` is that on purpose?

**Comment by zzo** [comment]
ya that is NOT intended not sure why that is happening - I'll fix before moving on

**Comment by zzo** [comment]
ya ok figured it out (but not after accidentally publishing a 2.0 release of karma-junit-reporter which I have fixed :)  carrying on...

**Comment by zzo** [comment]
ok all of the plugins are updated to 1.0.0 (or 1.0.1 :))  
Last thing on the list is clean up issues

**Comment by dignifiedquire** [comment]
It's done :)

**Comment by zzo** [comment]
PROFIT!

**Comment by zzo** [comment]
do need to mention we can now go back and clean up module deps that have the '1.0 || xxxx' syntax - this can be done as plugins &c actually need to be updated
